src/background_music/FrameBasedEffectPlugin.h

```diff
@@ -0,0 +1,190 @@
+#ifndef AUDACIOUS_PLUGINS_FRAMEBASEDEFFECTPLUGIN_H
+#define AUDACIOUS_PLUGINS_FRAMEBASEDEFFECTPLUGIN_H
+/*
+ * audacious-plugins/FrameBasedEffectPlugin.h
+ *
+ * Added by michel on 2023-09-23
+ * Copyright (C) 2015-2023 Michel Fleur.
+ * Source https://github.com/emmef/audacious-plugins
+ * Email audacious-plugins@emmef.org
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include <libaudcore/plugin.h>
+#include <libaudcore/ringbuf.h>
+
+#include "Detection.h"
+
+template<class D>
+class FrameBasedEffectPlugin : public EffectPlugin
+{
+    static_assert(std::is_base_of_v<Detection, D>);
+
+public:
+    constexpr FrameBasedEffectPlugin(const PluginInfo info, int order)
+        : EffectPlugin(info, order, true)
+    {
+    }
+    virtual ~FrameBasedEffectPlugin() = default;
+
+    bool init() final;
+    void cleanup() final;
+
+    void start(int & channels, int & rate) final;
+    Index<float> & process(Index<float> & data) final;
+    bool flush(bool force) final;
+    Index<float> & finish(Index<float> & data, bool end_of_playlist) final;
+    int adjust_delay(int delay) final;
+    bool offer_frame_return_if_output();
+
+    virtual bool after_init() { return true; }
+    virtual void before_cleanup() {}
+
+    Index<float> frame_in;
+    RingBuf<float> read_ahead_buffer;
+    Index<float> frame_out;
+    Index<float> output;
+    int processed_frames = 0;
+    int current_channels = 0, current_rate = 0, channel_last_read = 0;
+    D * detection = nullptr;
+};
+
+template<class D>
+bool FrameBasedEffectPlugin<D>::init()
+{
+    if (!detection)
+    {
+        detection = new D;
+    }
+    return after_init();
```

Comment:
Calling into the subclass here (with the virtual `after_init()` function) goes against the usual design pattern and is unnecessary. What you should do instead is add an overriding `init()` function to your subclass (BackgroundMusicPlugin). Make `BackgroundMusicPlugin::init` first call `FrameBasedEffectPlugin::init` and then perform whatever other initialization it needs to do.